kratos/spaces/ublas_space.h

```diff
@@ -723,18 +729,29 @@ class UblasSpace
         IndexPartition(system_size).for_each([&](std::size_t Index){
             bool empty = true;
 
-            const std::size_t col_begin = Arow_indices[Index];
-            const std::size_t col_end = Arow_indices[Index + 1];
+            const std::size_t col_begin = r_Arow_indices[Index];
+            const std::size_t col_end = r_Arow_indices[Index + 1];
 
             for (std::size_t j = col_begin; j < col_end; ++j) {
-                if(std::abs(Avalues[j]) > zero_tolerance) {
+                if(std::abs(r_Avalues[j]) > zero_tolerance) {
                     empty = false;
                     break;
                 }
             }
 
             if(empty) {
-                rA(Index, Index) = scale_factor;
+                const auto& it_Acol_indices_row_begin = r_Acol_indices.begin() + col_begin;
+                const auto& it_Acol_indices_row_end = r_Acol_indices.begin() + col_end;
+                const auto pos = std::lower_bound(it_Acol_indices_row_begin, it_Acol_indices_row_end, Index);
+                if (pos != it_Acol_indices_row_end) { // It was found
```

Comment:
This is incorrect. `std::lower_bound` finds the largest item less than or equal to `Index`, so if `Index` is not in the list, it won't generally go all the way to the end of the range.

You need an `std::upper_bound` as well and compare the two iterators. If they're identical, `Index` is not in the range.

Reply:
[example on godbolt](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:55,endLineNumber:7,positionColumn:55,positionLineNumber:7,selectionStartColumn:55,selectionStartLineNumber:7,startColumn:55,startLineNumber:7),source:'%23include+%3Ciostream%3E%0A%23include+%3Calgorithm%3E%0A%23include+%3Cvector%3E%0A%0Aint+main()%0A%7B%0A++++const+std::vector%3Cint%3E+column_indices+%3D+%7B1,+2,+3,+5,+6,+7%7D%3B%0A++++const+auto+it+%3D+std::lower_bound(column_indices.begin(),%0A+++++++++++++++++++++++++++++++++++++column_indices.end(),%0A+++++++++++++++++++++++++++++++++++++4)%3B%0A++++std::cout+%3C%3C+std::distance(column_indices.begin(),+it)+%3C%3C+std::endl%3B%0A%7D'),l:'5',n:'1',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',m:100,n:'0',o:'',s:0,t:'0'),(g:!((h:executor,i:(argsPanelShown:'1',compilationPanelShown:'0',compiler:g132,compilerName:'',compilerOutShown:'0',execArgs:'',execStdin:'',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',overrides:!(),runtimeTools:!(),source:1,stdinPanelShown:'1',wrap:'1'),l:'5',n:'0',o:'Executor+x86-64+gcc+13.2+(C%2B%2B,+Editor+%231)',t:'0')),header:(),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

Reply:
Done